### PR TITLE
Enable AutoFDO.

### PR DESCRIPTION
**Pull request by mikebenfield**
This largely involves implementing the options debug-info-for-profiling
and profile-sample-use and forwarding them on to LLVM.

AutoFDO can be used on x86-64 Linux like this:
rustc -O -Clink-arg='Wl,--no-rosegment' -Cdebug-info-for-profiling main.rs -o main
perf record -b ./main
create_llvm_prof --binary=main --out=code.prof
rustc -O -Cprofile-sample-use=code.prof main.rs -o main2

Now `main2` will have feedback directed optimization applied to it.

The create_llvm_prof tool can be obtained from this github repository:
https://github.com/google/autofdo

The option -Clink-arg='Wl,--no-rosegment' is necessary to avoid lld
putting an extra RO segment before the executable code, which would make
the binary silently incompatible with create_llvm_prof.

### CODE REVIEW

**Comment by mikebenfield** [comment]
I ran clang-format on the diff in `PassWrapper.cpp`, which reordered the `include`s. I can revert that part if desired.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 1]
```diff
@@ -740,19 +754,16 @@ struct LLVMRustSanitizerOptions {
...
-    LLVMRustSanitizerOptions *SanitizerOptions,
-    const char *PGOGenPath, const char *PGOUsePath,
-    bool InstrumentCoverage, bool InstrumentGCOV,
-    void* LlvmSelfProfiler,
+    bool MergeFunctions, bool UnrollLoops, bool SLPVectorize,
+    bool LoopVectorize, bool DisableSimplifyLibCalls, bool EmitLifetimeMarkers,
+    LLVMRustSanitizerOptions *SanitizerOptions, const char *PGOGenPath,
+    const char *PGOUsePath, bool InstrumentCoverage, bool InstrumentGCOV,
+    const char *PGOSampleUsePath, bool DebugInfoForProfiling,
+    void *LlvmSelfProfiler,
```
Comment (bjorn3) [inline compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:766]: Could you please avoid reformatting this file?

**Comment by bjorn3** [comment]
> The option -Clink-arg='Wl,--no-rosegment' is necessary to avoid lld putting an extra RO segment before the executable code, which would make the binary silently incompatible with create_llvm_prof.

This will make read-only data executable, which depending on what program you have may or may not be a bad idea.

**Comment by mikebenfield** [comment]
> This will make read-only data executable, which depending on what program you have may or may not be a bad idea.

Indeed. Unfortunately it's the only way to deal with a current limitation of the create_llvm_prof tool, which can't deal with that RO segment. I've been told they plan to remove that limitation eventually, but that it won't happen in the immediate future.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 1 above)
Comment (mikebenfield) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Fixed.

**Comment by mikebenfield** [comment]
Rebased, undid formatting in PassWrapper.cpp, fixed formatting in config.rs.

**Comment by mikebenfield** [comment]
And fixed the commit message.

**Comment by cjgillot** [comment]
I don't know anything about LLVM.
r? @nikic

**Comment by mikebenfield** [comment]
A fix is in the works for the `create_llvm_prof` tool so it will no longer require `--no-rosegment`, so I removed that from the commit message.

**Comment by mikebenfield** [comment]
@nikic Would you mind taking a look at this?

**Review by nikic — COMMENTED** [review]
This looks pretty straightforward and seems to mirror what clang does. Some high-level notes:

 * New options are always introduced as unstable `-Z` options first, and then stabilized as `-C` options. You should switch the new options to use `-Z` and open a tracking issue for this feature.
 * Not strictly required, but adding some basic docs on how to use this to `src/doc/unstable-book/src/compiler-flags` wouldn't hurt. You'll find `profile.md` and `instrument_coverage.md` there for related functionality. This could basically just be what you have in your PR description, maybe with an introduction on what this is actually about (it's easy to get lost in all these profiling, coverage and instrumentation options...)
 * It would be great to have some kind of testing for this. I see that the dependence on `create_llvm_prof` would make an end-to-end test inconvenient. If nothing else, we can test that `-Cdebuginfo-for-profiling` adds the function attribute using a test in src/test/codegen.

File: compiler/rustc_codegen_llvm/src/callee.rs [hunk 2]
```diff
@@ -84,6 +84,8 @@ pub fn get_fn(cx: &CodegenCx<'ll, 'tcx>, instance: Instance<'tcx>) -> &'ll Value
 
         attributes::from_fn_attrs(cx, llfn, instance);
 
+        attributes::apply_use_sample_profile(cx.sess(), llfn);
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_codegen_llvm/src/callee.rs]: Done.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 3]
```diff
@@ -516,21 +518,33 @@ extern "C" void LLVMRustDisposeTargetMachine(LLVMTargetMachineRef TM) {
...
+  }
+
+  auto AddAddDiscriminatorsPass = [](const PassManagerBuilder &Builder,
+                                     PassManagerBase &PM) {
+    PM.add(llvm::createAddDiscriminatorsPass());
+  };
+
+  if (DebugInfoForProfiling || PGOSampleUsePath) {
+    unwrap(PMBR)->addExtension(PassManagerBuilder::EP_EarlyAsPossible,
+                               AddAddDiscriminatorsPass);
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Done.

File: src/bootstrap/compile.rs [hunk 4]
```diff
@@ -614,6 +614,18 @@ impl Step for Rustc {
...
+            if compiler.stage == 1 {
+                cargo.rustflag(&format!("-Cprofile-sample-use={}", path));
+            }
+        }
+
+        if builder.config.rust_debug_info_for_profiling {
+            if compiler.stage == 1 {
+                cargo.rustflag("-Cdebug-info-for-profiling");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply src/bootstrap/compile.rs]: Removed this bootstrapping stuff.

File: compiler/rustc_session/src/options.rs [hunk 5]
```diff
@@ -967,6 +969,8 @@ options! {
     profile_generate: SwitchWithOptPath = (SwitchWithOptPath::Disabled,
         parse_switch_with_opt_path, [TRACKED],
         "compile the program with profiling instrumentation"),
+    profile_sample_use: Option<PathBuf> = (None, parse_opt_pathbuf, [TRACKED],
+        "use the given `.prof` file for sampled profile-guided optimization (also known as AutoFDO)"),
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_session/src/options.rs]: Done.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 6]
```diff
@@ -790,11 +805,19 @@ LLVMRustOptimizeWithNewPassManager(
...
   if (PGOGenPath) {
-    assert(!PGOUsePath);
-    PGOOpt = PGOOptions(PGOGenPath, "", "", PGOOptions::IRInstr);
+    assert(!PGOUsePath && !PGOSampleUsePath);
+    PGOOpt = PGOOptions(PGOGenPath, "", "", PGOOptions::IRInstr,
+                        PGOOptions::NoCSAction, DebugInfoForProfiling);
   } else if (PGOUsePath) {
-    assert(!PGOGenPath);
-    PGOOpt = PGOOptions(PGOUsePath, "", "", PGOOptions::IRUse);
+    assert(!PGOSampleUsePath);
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:801]: If `PGOGenPath` tested true we'd be in the first branch of the `if`.

**Comment by mikebenfield** [comment]
I made these into `-Z` options and added some documentation.

Testing seems more difficult. `debug-info-for-profiling` actually doesn't add a function attribute. `profile-sample-use` does, but it requires a profile, and I'm not sure whether it's wise to include a profile. Please let me know what you think.

**Comment by mikebenfield** [comment]
@nikic Mind taking another look?

File: compiler/rustc_codegen_ssa/src/back/write.rs [hunk 7]
```diff
@@ -176,6 +178,8 @@ impl ModuleConfig {
                 SwitchWithOptPath::Disabled
             ),
             pgo_use: if_regular!(sess.opts.cg.profile_use.clone(), None),
+            pgo_sample_use: if_regular!(sess.opts.cg.profile_use.clone(), None),
```
Comment (nikic) [inline compiler/rustc_codegen_ssa/src/back/write.rs:181]: Shouldn't this be assigning `profile_sample_use`?

File: src/bootstrap/config.rs [hunk 8]
```diff
@@ -516,6 +517,7 @@ struct Rust {
     new_symbol_mangling: Option<bool>,
     profile_generate: Option<String>,
     profile_use: Option<String>,
+    profile_sample_use: Option<String>,
```
Comment (nikic) [inline src/bootstrap/config.rs:520]: The changes to this file should no longer be necessary, as the rest of the bootstrap support was dropped.

**Comment by mikebenfield** [comment]
I addressed both recent comments.

**Review by nikic — APPROVED** [review]
Looks good, can you please squash the commits?

**Comment by mikebenfield** [comment]
Squashed.

**Comment by mikebenfield** [comment]
And fixed the commit message.

**Pull request merged**